core-lib/Streams.ns

```diff
@@ -0,0 +1,1037 @@
+	%responsibility put:
+			Subclasses must extend the put: method to do the actual work of writing the element.
+	%responsibility size:
+			Optional - Any subclass that wishes to start off with readable elements in the stream must send
+			the message size: to indicate how many there are.
+	*)| size <Integer> ::= 0.  |) (
+
+		public contentsWritten  ^ <Vector[E]> = (
+			| cltn <Vector[E]>  origPos <Integer> |
+			cltn:: self newCollectionOfSize: self position.
+			origPos:: self position.
+			self position: 0.
+			1 to: origPos do: [ :i <Integer> |
+				cltn at: i put: self next.	].
+			self position: origPos.
+			^cltn
+		)
+
+		public put: el <E> ^ <E> = (
+			(* Subclasses must extend this to do the actual writing of the element.  This impl just keeps track of bookkeeping. *)
+			| pos <Integer> |
+			self position: (pos:: self position + 1).
+			pos > self size
+				ifTrue: [ self size: pos ].
+			^el
+		)
+	) : (
+		on: sc <Vector[E]>  ^ <ReadWriteStream[E]> = (
+			(* Create a new read write stream on a sequenceable collection, assuming that it contains no readable elements yet *)
+			^self on: sc numElements: 0
+		)
+
+		on: sc <Vector[E]> numElements: n <Integer> ^ <ReadWriteStream[E]> = (
+			(* Create a new read write stream on a sequenceable collection, assuming that it contains n readable elements to start *)
+			^SeqCltnReadWriteStream on: sc numElements: n
+		)
+	)
+
+	class CharacterInput = BasicInputStream(
+	(* This mixin adds character specific messages to an InputStream class. *)) (
+		contentsRemaining ^ <String> = (
+			| con <Vector[Character]> |
+			con:: super contentsRemaining.
+
+			^con isString
+				ifTrue: [ con ]
+				ifFalse: [ String from: (con asArray) ]
+		)
+
+		newCollectionOfSize: n <Integer> ^ <String> = (
+			^Array new: n
+		)
+
+		public next: nElements <Integer> ^ <String> = (
+			| con <Vector[Character]> |
+			con:: super next: nElements.
+			^ String from: con.
+		)
+
+		upTo: o <Object> ^ <String> = (
+			| con <Vector[Character]> |
+			con:: super upTo: o.
+			^con isString
+				ifTrue: [ con ]
+				ifFalse: [ String from: (con asArray) ]
+		)
+	) : (
+	)
+
+	public class CharacterInputConverter = CharacterInput (
+	(* Instances are used to create a CharInputStream that wraps a byte input stream*)
+	| binaryStream <STRM> |) (
+		actualAtEnd ^ <Boolean> = (
+			^self binaryStream atEnd
+		)
+
+		actualNext ^ <Character> = (
+				| c <Character> |
+			c:: Character from: self binaryStream next.
+			(usingCrLfConvention and: [ c = '\r' ])
+						ifTrue: [ self binaryStream atEnd
+								ifFalse: [ self binaryStream peek = (Character value: '\n')
+													ifTrue: [  self binaryStream next ]	]].
+			^c
+		)
+
+		public close = (
+			self binaryStream close.
+		)
+
+		public size = (
+			^binaryStream size.
+		)
+
+		public initBinaryStream: binaryStrm <STRM> = (
+			self binaryStream: binaryStrm
+		)
+
+		upTo: del <Object> ^ <Str> = (
```

Comment:
Str?

Reply:
I think we want a basic name check for type annotations.
To make sure that we at least use things that exist :D